src/vs/editor/common/model/textModelTokens.ts

```diff
@@ -208,102 +221,74 @@ export class TextModelTokenization extends Disposable {
+		this._defaultBackgroundTokenizer?.handleChanges();
 	}
 
 	public handleDidChangeAttached(): void {
-		this._beginBackgroundTokenization();
+		this._defaultBackgroundTokenizer?.handleChanges();
 	}
 
 	public handleDidChangeLanguage(e: IModelLanguageChangedEvent): void {
 		this._resetTokenizationState();
 		this._tokenizationPart.clearTokens();
 	}
 
-	//#endregion
-
 	private _resetTokenizationState(): void {
 		const [tokenizationSupport, initialState] = initializeTokenization(this._textModel, this._tokenizationPart);
 		if (tokenizationSupport && initialState) {
 			this._tokenizationStateStore = new TokenizationStateStore(tokenizationSupport, initialState);
 		} else {
 			this._tokenizationStateStore = null;
 		}
-		this._beginBackgroundTokenization();
-	}
-
-	private _isScheduled = false;
-	private _beginBackgroundTokenization(): void {
-		if (this._isScheduled || !this._textModel.isAttachedToEditor() || !this._hasLinesToTokenize()) {
-			return;
-		}
-
-		this._isScheduled = true;
-		runWhenIdle((deadline) => {
-			this._isScheduled = false;
-
-			this._backgroundTokenizeWithDeadline(deadline);
-		});
-	}
-
-	/**
-	 * Tokenize until the deadline occurs, but try to yield every 1-2ms.
-	 */
-	private _backgroundTokenizeWithDeadline(deadline: IdleDeadline): void {
-		// Read the time remaining from the `deadline` immediately because it is unclear
-		// if the `deadline` object will be valid after execution leaves this function.
-		const endTime = Date.now() + deadline.timeRemaining();
-
-		const execute = () => {
-			if (this._isDisposed || !this._textModel.isAttachedToEditor() || !this._hasLinesToTokenize()) {
-				// disposed in the meantime or detached or finished
-				return;
-			}
 
-			this._backgroundTokenizeForAtLeast1ms();
+		this.backgroundTokenizer.clear();
 
-			if (Date.now() < endTime) {
-				// There is still time before reaching the deadline, so yield to the browser and then
-				// continue execution
-				setTimeout0(execute);
-			} else {
-				// The deadline has been reached, so schedule a new idle callback if necessary
-				this._beginBackgroundTokenization();
+		this._defaultBackgroundTokenizer = null;
+		if (this._tokenizationStateStore) {
+			const b: IBackgroundTokenizationStore = {
+				setTokens: (tokens) => {
+					this._tokenizationPart.setTokens(tokens);
+				},
+				backgroundTokenizationFinished: () => {
+					this._tokenizationPart.handleBackgroundTokenizationFinished();
+				},
+				setEndState: (lineNumber, state) => {
+					if (!state) {
+						throw new BugIndicatingError();
+					}
+					const invalidLineStartIndex = this._tokenizationStateStore?.invalidLineStartIndex;
+					if (invalidLineStartIndex !== undefined && lineNumber - 1 >= invalidLineStartIndex) {
+						// Don't accept states for definitely valid states
+						this._tokenizationStateStore?.setEndState(this._textModel.getLineCount(), lineNumber - 1, state);
+					}
+				},
+			};
+
+			this.backgroundTokenizer.clear();
```

Comment:
This is not necessary because `this.backgroundTokenizer.clear();` is called above